First PR - Failover Connector skeleton (#28818)
open-telemetry/opentelemetry-collector-contrib: connector/failoverconnector/README.md

| logs | logs | [In development]           |

[Exporter Pipeline Type]: https://github.com/open-telemetry/opentelemetry-collector/blob/main/connector/README.md#exporter-pipeline-type
[Receiver Pipeline Type]: https://github.com/open-telemetry/opentelemetry-collector/blob/main/connector/README.md#receiver-pipeline-type
[Stability Level]: https://github.com/open-telemetry/opentelemetry-collector#stability-levels
<!-- end autogenerated section -->

Allows for health based routing between trace, metric, and log pipelines depending on the health of target downstream exporters.

## Configuration

If you are not already familiar with connectors, you may find it helpful to first visit the [Connectors README].

The following settings are available:

- `priority (required)`: list of pipeline level priorities in a 1 - n configuration, multiple pipelines can sit at a single priority level.
- `retry_interval (required)`: the frequency at which the pipeline levels will attempt to reestablish connection with all higher priority levels
- `retry_gap (required)`: the amount of time between trying two separate priority levels in a single retry_interval timeframe

> - `max_retry (required)`: the maximum retries per level

Reviewer: Typically we try to provide reasonable defaults wherever possible, in order to make it as simple as possible to get started with the component. These fields all seem like they could have reasonable defaults.

Author: The idea is that it should retry a maximum of one unhealthy level before returning to the stable level to avoid back pressure through the collector. Especially with retries enabled on the exporters it could take several seconds for each level that is retried. If we’re on a low enough level and each higher priority level needs to be retried prior to returning to stability, that timeframe could become significant very quickly.

Example:

```yaml
connectors:
  failover:

>     priority:

Reviewer: We've got a mix of terms throughout this doc. "priority", "priority level" and "level". IMO "priority level" is the most clear. Maybe?
```suggestion
    priority_levels:
```

Author: Agreed

>       - [traces/first, traces/fourth]

Reviewer: Since these are at the same priority level, I wonder if it would be a bit clearer to new users if we renamed "fourth" to "also_first" or anything to that effect. It seems like the current name implies that it's somehow treated as a different priority.

      - [traces/second]
      - [traces/third]

>     retry_pipeline_interval: 5m

Reviewer: This seems sufficiently descriptive to me. WDYT?
```suggestion
    retry_interval: 5m
```

Author: Agreed

>     retry_gap: 1m

Reviewer: Is "interval" accurate for how you expect this timing to behave? We use this term in a lot of similar situations, so I'd suggest it here as well unless there's a nuance to the behavior which makes it inaccurate.
```suggestion
    retry_interval: 1m
```

Author: So the interval is the time between the cycles of starting over from the highest priority pipeline down to the current stable one while gap is how long it waits between two levels in a single interval’s timeframe. If there’s a naming convention that aligns better to other similar uses, I’m all for it.

Reviewer: Hmm, what's the reason for needing to add a delay here? Why not just immediately sent to the next priority after we drop down a level?

Author: The idea is that it should retry a maximum of one unhealthy level before returning to the stable level to avoid back pressure through the collector. Especially with retries enabled on the exporters it could take several seconds for each level that is retried. If we’re on a low enough level and each higher priority level needs to be retried prior to returning to stability, that timeframe could become significant very quickly.

Author: @djaglowski, any thoughts?

Reviewer: Ok, if I'm understanding correctly, any level which fails a retry will immediately end all retries until the next interval. However, if a retry is successful, we will both 1) reset the current level to the successful level, and 2) continue to retry any lower levels - and this gap is the time we wait until we retry the next level?

Author: @djaglowski  Let me give an example:

connectors:
  failover:
    priority:
      - [traces/first]
      - [traces/second]
      - [traces/third]
      - [traces/fourth]
    retry_interval: 10m
    retry_gap: 1m
    max_retry: 10

Lets say we are currently on level 4 on the `priority_level` list. At the start of the retry_interval, the connector will try to reestablish the pipeline on level 1 (trace/first). If it fails, the connector will return to level 4 (traces/fourth) and wait the 1m as the retry_gap, when that 1m passes it will now retry level 2 (traces/second) and if that fails will first return to level 4 before waiting another 1m until trying level 3. Once it tries level 3 and it fails, it will return to level 4 and wait the 10m retry_interval again before repeating the process.

The idea for the retry_gap and returning to the _stable level_ between the _retry level_ is to make sure that the connector has a chance to export data without introducing a significant time gap or significant backpressure.

The idea for setting a retry_interval is to avoid constant retries, and allow for whatever needs to be resolved downstream to be resolved (e.g., a collector gateway that is down)

To answer you other question, if a retry is successful then yes the retried level becomes the stable level, and the connector will continue to retry any lower levels that haven't exceeded the `max_retries`

Reviewer: Thanks for clarifying @akats7. It seems worth capturing more detail in the readme description. Perhaps this example would be useful as a standalone section and then you can refer to it from the short descriptions in the table.

Author: Sure thing, will add that to the readme.

>     max_retry: 10

Reviewer: This is documented below as "max_retries", which I think is clearer.
```suggestion
    max_retries: 10
```

Author: Addressed

service:
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [failover]
    traces/first:
      receivers: [failover]
      exporters: [otlp/first]
    traces/second:
      receivers: [failover]
      exporters: [otlp/second]
    traces/third:
      receivers: [failover]
      exporters: [otlp/third]
    traces/fourth:
      receivers: [failover]
      exporters: [otlp/fourth]
```

> The connector intakes a list of 'levels' each of which can contain multiple pipelines. If any pipeline at a stable level fails, the level is considered unhealthy and the connector will move down one priority level and route all data to the new level (assuming it is stable).

Reviewer: Where possible, it's helpful to use the exact config keys so the user doesn't have to assume what is intended.
```suggestion
The connector intakes a list of `priority_levels` each of which can contain multiple pipelines. If any pipeline at a stable level fails, the level is considered unhealthy and the connector will move down one priority level and route all data to the new level (assuming it is stable).
```

Author: Addressed

> The connector will periodically try to reestablish a stable connection with the higher priority levels. RetryInterval will be the frequency at which the connector will try to iterate through all unhealthy higher priority levels while retryGap is how long it will wait after a failed retry at one level before retrying the next level (if retry gap is 2m, after trying to reestablish level 1, it will wait 2m before trying level 2) It will retry a maximum of one unhealthy level before returning to the current stable level.)

Reviewer: ```suggestion
The connector will periodically try to reestablish a stable connection with the higher priority levels. `retry_interval` will be the frequency at which the connector will try to iterate through all unhealthy higher priority levels while `retry_gap` is how long it will wait after a failed retry at one level before retrying the next level (if retry gap is 2m, after trying to reestablish level 1, it will wait 2m before trying level 2) It will retry a maximum of one unhealthy level before returning to the current stable level.)
```

Author: Addressed

> There is a max_retries config param as well that will track how many retries have occurred at each level, and once the max is hit, it will no longer retry that priority level.

Reviewer: ```suggestion
There is a `max_retries` config param as well that will track how many retries have occurred at each level, and once the max is hit, it will no longer retry that priority level.
```

Author: Addressed

There is a `max_retries` config param as well that will track how many retries have occurred at each level, and once the max is hit, it will no longer retry that priority level.

#### Example with Explanation:

```yaml
connectors:
  failover:
    priority_levels:
      - [traces/first]
      - [traces/second]
      - [traces/third]
      - [traces/fourth]
    retry_interval: 5m
    retry_gap: 1m
    max_retries: 10
```

Assume the current stable level is level 4 (traces/fourth) on the priority_level list. 
At the start of the `retry_interval`, the connector will try to reestablish the pipeline on level 1 (trace/first). If it fails, the connector will return to level 4 (traces/fourth) and wait the 1m as the `retry_gap`, when that 1m passes it will now retry level 2 (traces/second) and if that fails will first return to level 4 before waiting another 1m until trying level 3. 

> Once it tries level 3 and it fails, it will return to level 4 and wait the 10m retry_interval again before repeating the process. If a retry is successful then the retried level becomes the stable level, and the connector will continue to retry any lower levels that haven't exceeded the `max_retries`.

Reviewer 2: > If a retry is successful then the retried level becomes the stable level, and the connector will continue to retry any lower levels that haven't exceeded the `max_retries`.

I did not get the meaning of the `lower levels`. You saild `max_retries` denotes "the maximum retries per level". Why is this changed to lower levels here?
